### PR TITLE
Implement a cgi alive test when determining what VSO endpoint to use

### LINKED ISSUE #6361: When deciding which VSO endpoint to connect to we should determine if the soap address in the WDSL is accessible
By default we access the WDSL at http://docs.virtualsolar.org/WSDL/VSOi_rpc_literal.wsdl but the soap endpoint is http://vso.nso.edu/cgi-bin/VSO/PROD/vsoi_wsdl.cgi we should test both somehow

### PR DESCRIPTION
**Pull request by Cadair**
fixes #6361 

I have no idea how to test this, as we are testing that a url in a remote file is accessible. I guess the only way would be to trick the function into downloading a pre-faked WSDL file.

### CODE REVIEW

File: sunpy/net/vso/vso.py [hunk 1]
```diff
@@ -61,12 +61,37 @@ def check_connection(url):
...
+    This is weird enough that it probably satisfies us for this check.
+    """
+    try:
+        return urlopen(url).getcode() == 411
+    except HTTPError as e:
+        if e.code == 411:
+            return True
+        warn_user(f"Connection to {url} failed with error {e}. Retrying with different url and port.")
+        return None
+    except (socket.error, socket.timeout, HTTPError, URLError) as e:
```
Comment (Cadair) [inline sunpy/net/vso/vso.py:77]:
```suggestion
    except (socket.error, socket.timeout, URLError) as e:
```

File: sunpy/net/vso/vso.py [hunk 2]
```diff
@@ -61,12 +61,37 @@ def check_connection(url):
...
 
 
+def check_cgi_connection(url):
+    """
+    At the moment there is no way to make an "are you alive" request to the
+    cgi, so we just hit it with a HTTP get and it gives us back a 411 response.
+    This is weird enough that it probably satisfies us for this check.
+    """
+    try:
+        return urlopen(url).getcode() == 411
```
Comment (alasdairwilson) [inline sunpy/net/vso/vso.py:71]:
```suggestion
        return urlopen(url, timeout=10).getcode() == 411
```

We probably don't want to wait the 5 minute default timeout?

**Comment by alasdairwilson** [comment]
Thanks for this, very edge case but frustrating to encounter. 👍

File: sunpy/net/vso/vso.py [hunk 3]
```diff
@@ -40,33 +40,63 @@
...
+    At the moment there is no way to make an "are you alive" request to the
+    cgi, so we just hit it with a HTTP get and it gives us back a 411 response.
+    This is weird enough that it probably satisfies us for this check.
+    """
+    try:
+        return urlopen(url, timeout=15).getcode() == 411
+    except HTTPError as e:
+        if e.code == 411:
+            return True
+        warn_user(f"Connection to {url} failed with error {e}. Retrying with different url and port.")
```
Comment (dstansby) [inline sunpy/net/vso/vso.py:80]: I think this error message is outside the scope of this function, as whatever calls it could do something else instead of retyring with a different url and port?

File: sunpy/net/vso/vso.py [hunk 4]
```diff
@@ -40,33 +40,63 @@
...
+        return None
+    except (socket.error, socket.timeout, URLError) as e:
+        warn_user(f"Connection to {url} failed with error {e}. Retrying with different url and port.")
```
Comment (dstansby) [inline sunpy/net/vso/vso.py:83]: Similarly for this message

File: sunpy/net/vso/vso.py [hunk 5]
```diff
@@ -40,33 +40,63 @@
...
-        return urlopen(url).getcode() == 200
+        return urlopen(url, timeout=15).getcode() == 200
     except (socket.error, socket.timeout, HTTPError, URLError) as e:
         warn_user(f"Connection to {url} failed with error {e}. Retrying with different url and port.")
         return None
 
 
+def check_cgi_connection(url):
+    """
+    At the moment there is no way to make an "are you alive" request to the
```
Comment (dstansby) [inline sunpy/net/vso/vso.py:71]: Could you quickly document in the docstring what returning `None` means here, as opposed to `True` or `False`?

**Review by dstansby — CHANGES_REQUESTED** [review]

File: sunpy/net/vso/vso.py (hunk 3 above)
Comment (Cadair) [reply sunpy/net/vso/vso.py:80]: These messages are here because they need the exception context to be able to display the error to the user (as the exception is not raised).

Comment (dstansby) [reply sunpy/net/vso/vso.py:80]: Doesn't returning `False` provide the context for calling code to raise these warning messages instead (if they are indeed retrying with a different URL)?

File: sunpy/net/vso/vso.py (hunk 5 above)
Comment (Cadair) [reply sunpy/net/vso/vso.py:71]: I made it return False, I was just copying the previous implementation, but anything Falsey works.

**Comment by dstansby** [comment]
Nudge - I'm happy to try and finish this up @Cadair if you don't have time?

File: sunpy/net/vso/vso.py (hunk 3 above)
Comment (Cadair) [reply sunpy/net/vso/vso.py:80]: The calling context wouldn't have the exception instance which has the reason why that URL erred out.

**Comment by nabobalis** [comment]
I wasn't sure what was left for this but I rebased and added some mock tests for the new functions.

I think a better mocked test would be to make the first urlopen fail then see if it connects to a second URL later on as a real remote test.

**Comment by Cadair** [comment]
The online test fails are unrelated

File: sunpy/net/vso/tests/test_vso.py [hunk 6]
```diff
@@ -9,17 +14,33 @@
...
 
 
 class MockQRRecord:
     """
     Used to test sunpy.net.vso.QueryResponse.build_table(...)
     """
-    def __new__(cls, start_time=None, end_time=None, size=0, source='SOHO', instrument='aia',
-                extent=None, fileid="spam"):
+    def __new__(
+        cls,
```
Comment (nabobalis) [inline sunpy/net/vso/tests/test_vso.py:35]: What did this?

Comment (Cadair) [reply sunpy/net/vso/tests/test_vso.py:35]: I did

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
LGTM, shall we merge?
